timeseries/src/autogluon/timeseries/models/tabular/tabular_model.py

```diff
@@ -0,0 +1,198 @@
+
+        return pd.concat(dataframe_per_item, ignore_index=True)
+
+    def _generate_features_from_freq(self, freq: str):
+        self._lag_indices = get_lags_for_frequency(freq)
+        self._time_features = time_features_from_frequency_str(freq)
+
+    def _fit(
+        self,
+        train_data: TimeSeriesDataFrame,
+        val_data: Optional[TimeSeriesDataFrame] = None,
+        time_limit: int = None,
+        **kwargs,
+    ) -> None:
+        self._check_fit_params()
+        if self.tabular_predictor._learner.is_fit:
+            raise AssertionError(f"{self.name} predictor has already been fit!")
+        self._generate_features_from_freq(train_data.freq)
+
+        train_data, _ = self._normalize_targets(train_data)
+        train_df = self._get_features_dataframe(train_data)
+        # Remove features that are completely missing in the training set
+        train_df.dropna(axis=1, how="all", inplace=True)
+        self._available_features = train_df.columns
+
+        if val_data is not None:
+            if val_data.freq != train_data.freq:
+                raise ValueError(
+                    f"train_data and val_data must have the same freq (received {train_data.freq} and {val_data.freq})"
+                )
+            val_data, _ = self._normalize_targets(val_data)
+            val_df = self._get_features_dataframe(val_data, last_k_values=self.prediction_length)
+            val_df = val_df[self._available_features]
+        else:
+            logger.warning(
+                f"No val_data was provided to {self.name}. "
+                "TabularPredictor will generate a validation set without respecting the temporal ordering."
+            )
+            val_df = None
+
+        # TODO: Other presets for TabularPredictor?
+        # TODO: Speed up prediction for Quantile XT/RF models
+        # TODO: Add catboost with MultiQuantile loss (after catboost v1.1)
+        # TODO: Other tabular models in quantile? (LightGBM?)
+        self.tabular_predictor.fit(
+            train_data=train_df,
+            tuning_data=val_df,
+            time_limit=time_limit,
+            excluded_model_types=["XT", "RF"],
+            verbosity=2,
+        )
+
+    def predict(self, data: TimeSeriesDataFrame, quantile_levels: List[float] = None, **kwargs) -> TimeSeriesDataFrame:
+        if quantile_levels is not None:
+            raise ValueError(f"{self.name} cannot predict custom quantiles. Please set `quantile_levels=None`.")
+
+        data, scale_per_item = self._normalize_targets(data)
+
+        last_observed_timestamp = data.slice_by_timestep(-1, None).index.get_level_values(TIMESTAMP)
+        offset = pd.tseries.frequencies.get_offset(data.freq)
+        item_ids = data.item_ids
+        nan_array = np.full(len(item_ids), fill_value=np.nan)
+
+        predictions_list = []
+        full_df = data
+
+        # Autoregressively generate predictions
+        for idx in range(1, self.prediction_length + 1):
```

Comment:
this part looks like it depends on a lot of moving parts and indexing logic. can we add a test maybe mocking out `_get_features_dataframe` to make sure the features align with the value to be forecasted?

Reply:
Probably both, but I hope the latter is already tested.